File: RayTracer/Light/PointLight.cs
using RayTracer.Common;
using RayTracer.Material;
using RayTracer.Shape;
using RayTracer.Tracer;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using RayTracer.BVH;

namespace RayTracer.Lighting
{
    [Serializable]
    public class PointLight : Light
    {

        public PointLight(Point3 from, MyColor color)
        {
            Position = from;
            Color = color;

        }
        public PointLight(float[] param)
            : this(new Point3(param[0], param[1], param[2]), new MyColor(param[3], param[4], param[5]))
        {
        }

        public Point3 Position
        { get; set; }

        public override Vec3 GetPointToLight(Point3 point)
        {
            return new Vec3(point, Position);
        }
        
        public override bool IsEffective(Point3 point, Geometry geometry, Container bvh)
        {
            Vec3 pointToLight = GetPointToLight(point);
            if (pointToLight * geometry.GetNormal(point) < 0)
                return false;

            Ray shadowRay = new Ray(point, pointToLight.Normalize());

            if (bvh.IsIntersecting(shadowRay))
            {
                if (bvh.Geo != null)
                {
                    shadowRay.TransformInv(bvh.Geo.Trans);
                    if (bvh.Geo.IsIntersecting(shadowRay))
                        if (shadowRay.IntersectDistance < pointToLight.Magnitude)
                            return false;
                }
                else
                {
                    foreach (Container bin in bvh.Childs)
                        if (!IsEffective(point, geometry, bin)) return false;
                }
            }

            return true;
        }

        public override float GetAttValue(Point3 point, Attenuation attenuation)
        {
            if (!attenuation.Equals(new Attenuation()))
                return 1f;

            float d = GetPointToLight(point).Magnitude;
            return 1f /
                (attenuation.Constant +
                attenuation.Linear * d +
                attenuation.Quadratic * (float)Math.Pow(d, 2));

        }


    }
}

File: RayTracer - BVH - CUDA/RayTracerCUDA/PointLight.cpp
#include "PointLight.h"


PointLight::PointLight()
{
}


PointLight::~PointLight()
{
}

bool PointLight::isEffective(vec3 & point, std::shared_ptr< Container> bvh)
{
	vec3 pointToLight = getPointToLight(point);
	vec3 dir = pointToLight;
	dir = dir.Normalize();
	Ray shadowRay(point, dir);
	std::vector<std::shared_ptr< Container>> bins;

	bins.push_back(bvh);
	while (bins.size() > 0)
	{
		std::shared_ptr< Container> currBin = bins.back();
		if (currBin->isIntersecting(shadowRay))
		{
			if (currBin->geo != nullptr)
			{
				shadowRay.transInv(currBin->geo->getTrans());

				if (currBin->geo->isIntersecting(shadowRay))
					if (shadowRay.intersectDist < pointToLight.Magnitude())
						return false;
			}
			else
			{
				//if (!PointLight::isEffective(point, *bvh.lChild))
				//	return false;
				//if (!PointLight::isEffective(point, *bvh.rChild))
				//	return false;
				bins.push_back(currBin->rChild);
				bins.push_back(currBin->lChild);
				std::vector<std::shared_ptr <Container>>::iterator lPos = std::find(bins.begin(), bins.end(), currBin->lChild);
				std::vector<std::shared_ptr <Container>>::iterator currPos = std::find(bins.begin(), bins.end(), currBin);
				std::swap(*lPos, *currPos);

			}
		}
		bins.pop_back();
	}
	return true;
}

float PointLight::getAttValue(vec3 & point, Attenuation & att)
{
	float d = getPointToLight(vec3(point)).Magnitude();
	return 1.f / (att.cons + (att.line * d) + (att.line * d * d));
}

vec3 PointLight::getPointToLight(const vec3 & point)
{
	return vec3(point, (*pos));
}
